### PR TITLE
tensorsplit_op

### PR DESCRIPTION
**Pull request by lcylcy**

### CODE REVIEW

File: oneflow/core/functional/impl/math_functor.cpp [hunk 1]
```diff
@@ -1754,6 +1754,129 @@ class MovedimIntFunctor {
   }
 };
 
+class TensorSplitVecFunctor {
+ public:
+  TensorSplitVecFunctor() = default;
+  Maybe<TensorTuple> operator()(const std::shared_ptr<one::Tensor>& input, 
+                           const std::vector<int32_t>& indices_or_sections,
+                           const int32_t& dim) const {
```
Comment (wyushun) [inline oneflow/core/functional/impl/math_functor.cpp:1762]: 参考上个pr的comment：https://github.com/Oneflow-Inc/oneflow/pull/7275#discussion_r789443829

File: oneflow/core/functional/impl/math_functor.cpp [hunk 2]
```diff
@@ -1754,6 +1754,129 @@ class MovedimIntFunctor {
...
+    int32_t ndim = input->shape()->NumAxes();
+    CHECK_OR_RETURN((dim>=-ndim)&&(dim<ndim))<< "Dimension out of range (expected to be in range of ["
+                                              <<-ndim<<","<< ndim-1 <<"], but got "<<dim<<")";
+    int32_t pos_dim = dim>=0?dim:dim+ndim;
+    
+    std::vector<int64_t> start(ndim, 0);
+    std::vector<int64_t> stop(ndim);
+    std::vector<int64_t> step(ndim, 1);
+    for(int32_t i=0; i<ndim; i++){
+       stop[i] = input->shape()->At(i);
```
Comment (wyushun) [inline oneflow/core/functional/impl/math_functor.cpp:1772]: one::Tensor有个更简单的接口：`input->dim(i)`

reference：https://github.com/Oneflow-Inc/oneflow/blob/master/oneflow/core/framework/tensor.h#L49

File: oneflow/core/functional/impl/math_functor.cpp [hunk 3]
```diff
@@ -1754,6 +1754,129 @@ class MovedimIntFunctor {
...
+    
+    int32_t num_indices = indices_or_sections.size();
+    TensorTuple output(num_indices+1);
+    for(int32_t i = 0; i < num_indices; i++){
+       int32_t end_idx = indices_or_sections[i];
+       stop[pos_dim] = end_idx;
+       output[i] = JUST(Slice(input, start, stop, step));
+       start[pos_dim] = end_idx;
+    }
+    stop[pos_dim] = input->shape()->At(ndim-1);
```
Comment (wyushun) [inline oneflow/core/functional/impl/math_functor.cpp:1783]: 同上，https://github.com/Oneflow-Inc/oneflow/blob/master/oneflow/core/framework/tensor.h#L49

File: oneflow/core/functional/impl/math_functor.cpp [hunk 4]
```diff
@@ -1754,6 +1754,129 @@ class MovedimIntFunctor {
...
+    return output;
+  }
+};
+
+class TensorSplitIntFunctor {
+ public:
+  TensorSplitIntFunctor() = default;
+  Maybe<TensorTuple> operator()(const std::shared_ptr<one::Tensor>& input, 
+                                const int32_t& indices_or_sections,
+                                const int32_t& dim) const {
```
Comment (wyushun) [inline oneflow/core/functional/impl/math_functor.cpp:1795]: 同上：https://github.com/Oneflow-Inc/oneflow/pull/7258/files#r790376430

File: oneflow/core/functional/impl/math_functor.cpp [hunk 5]
```diff
@@ -1754,6 +1754,129 @@ class MovedimIntFunctor {
...
+    
+    const auto dim_size = input->shape()->At(pos_dim);
+    int64_t min_split_size = dim_size / indices_or_sections;
+    int64_t num_splits_one_extra = dim_size % indices_or_sections;
+    
+    std::vector<int64_t> start(ndim, 0);
+    std::vector<int64_t> stop(ndim);
+    std::vector<int64_t> step(ndim, 1);
+    for(int32_t i=0; i<ndim; i++){
+       stop[i] = input->shape()->At(i);
```
Comment (wyushun) [inline oneflow/core/functional/impl/math_functor.cpp:1810]: 同上，https://github.com/Oneflow-Inc/oneflow/pull/7258/files#r790378085

File: oneflow/core/functional/impl/math_functor.cpp [hunk 6]
```diff
@@ -1754,6 +1754,129 @@ class MovedimIntFunctor {
...
+    return output;
+  }
+};
+
+class HsplitIntFunctor {
+ public:
+  HsplitIntFunctor() = default;
+  Maybe<TensorTuple> operator()(const std::shared_ptr<one::Tensor>& input, 
+                                const int32_t& indices_or_sections) const {
+    int32_t ndim = input->shape()->NumAxes();
```
Comment (wyushun) [inline oneflow/core/functional/impl/math_functor.cpp:1831]: https://github.com/Oneflow-Inc/oneflow/blob/master/oneflow/core/framework/tensor.h#L51

File: oneflow/core/functional/impl/math_functor.cpp [hunk 7]
```diff
@@ -1754,6 +1754,129 @@ class MovedimIntFunctor {
...
+
+    return output;
+  }
+};
+
+class HsplitIntFunctor {
+ public:
+  HsplitIntFunctor() = default;
+  Maybe<TensorTuple> operator()(const std::shared_ptr<one::Tensor>& input, 
+                                const int32_t& indices_or_sections) const {
```
Comment (wyushun) [inline oneflow/core/functional/impl/math_functor.cpp:1830]: 同上：https://github.com/Oneflow-Inc/oneflow/pull/7258/files#r790376430

File: oneflow/core/functional/impl/math_functor.cpp [hunk 8]
```diff
@@ -1754,6 +1754,129 @@ class MovedimIntFunctor {
...
+    return TensorSplitInt(input, indices_or_sections, dim);
+  }
+};
+
+class HsplitVecFunctor {
+ public:
+  HsplitVecFunctor() = default;
+  Maybe<TensorTuple> operator()(const std::shared_ptr<one::Tensor>& input, 
+                                const std::vector<int32_t>& indices_or_sections) const {
+    int32_t ndim = input->shape()->NumAxes();
```
Comment (wyushun) [inline oneflow/core/functional/impl/math_functor.cpp:1847]: 同上：https://github.com/Oneflow-Inc/oneflow/pull/7258/files#r790379044

File: oneflow/core/functional/impl/math_functor.cpp [hunk 9]
```diff
@@ -1754,6 +1754,129 @@ class MovedimIntFunctor {
...
+    int32_t dim = (ndim == 1) ? 0 : 1;
+    return TensorSplitVec(input, indices_or_sections, dim);
+  }
+};
+
+class VsplitIntFunctor {
+ public:
+  VsplitIntFunctor() = default;
+  Maybe<TensorTuple> operator()(const std::shared_ptr<one::Tensor>& input, 
+                                const int32_t& indices_or_sections) const {
```
Comment (wyushun) [inline oneflow/core/functional/impl/math_functor.cpp:1858]: 同上：https://github.com/Oneflow-Inc/oneflow/pull/7258/files#r790379044

File: oneflow/core/functional/impl/math_functor.cpp [hunk 10]
```diff
@@ -1754,6 +1754,129 @@ class MovedimIntFunctor {
...
+    int32_t ndim = input->shape()->NumAxes();
+    CHECK_OR_RETURN(ndim>=2)<<"torch.vsplit requires a tensor with at least 2 dimension, but got a tensor with "<<ndim <<" dimensions!";
+    CHECK_OR_RETURN(indices_or_sections>0) << "indices_or_sections must greater than 0";
+    CHECK_OR_RETURN(input->shape()->At(0)% indices_or_sections == 0) << "torch.vsplit attempted to split along dimension " << 0
```
Comment (wyushun) [inline oneflow/core/functional/impl/math_functor.cpp:1862]: consider `input->dim()`

File: oneflow/core/functional/impl/math_functor.cpp [hunk 11]
```diff
@@ -1754,6 +1754,129 @@ class MovedimIntFunctor {
...
+    return TensorSplitInt(input, indices_or_sections, 0);
+  }
+};
+
+class VsplitVecFunctor {
+ public:
+  VsplitVecFunctor() = default;
+  Maybe<TensorTuple> operator()(const std::shared_ptr<one::Tensor>& input, 
+                                const std::vector<int32_t>& indices_or_sections) const {
+    int32_t ndim = input->shape()->NumAxes();
```
Comment (wyushun) [inline oneflow/core/functional/impl/math_functor.cpp:1874]: consider `input->ndim()`

File: python/oneflow/test/modules/test_hsplit.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+            dim1=random(3, 6),
+            dim2=random(3, 6),
+            dim3=random(3, 6),
+            dim4=random(3, 6),
+        ).to(device)
+        split = random(1, 3).to(int)
+        z = torch.hsplit(x, split)
+        return z[0]
+
+
```
Comment (wyushun) [inline python/oneflow/test/modules/test_hsplit.py:51]: 参考一下这个：https://github.com/Oneflow-Inc/oneflow/pull/7275#discussion_r789465928

File: python/oneflow/test/modules/test_tensor_split.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+            ndim=4,
+            dim1=random(3, 6),
+            dim2=random(3, 6),
+            dim3=random(3, 6),
+            dim4=random(3, 6),
+        ).to(device)
+        split = random(-3, 3).to(int)
+        dim = random(-3, 3).to(int)
+        z = torch.tensor_split(x, split,dim)
+        return z[0]
```
Comment (wyushun) [inline python/oneflow/test/modules/test_tensor_split.py:51]: 同上

File: python/oneflow/test/modules/test_vsplit.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        x = random_pytorch_tensor(
+            ndim=4,
+            dim1=random(3, 6),
+            dim2=random(3, 6),
+            dim3=random(3, 6),
+            dim4=random(3, 6),
+        ).to(device)
+        split = random(1, 3).to(int)
+        z = torch.vsplit(x, split)
+        return z[0]
```
Comment (wyushun) [inline python/oneflow/test/modules/test_vsplit.py:49]: 同上

**Review by wyushun — APPROVED** [review]
review done，写的太棒了，我写了一些comments，大多都和上个pr（https://github.com/Oneflow-Inc/oneflow/pull/7275）的雷同，你看情况自己酌情修改哈，我直接给你approve了，这样可以尽可能提高效率，你自己做足测试保证正确性就好～ @lcylcy

**Comment by lcylcy** [comment]
好的

**Review by luqiang-guo — APPROVED** [review]
